src/components/tl/ucp/reduce_scatter/reduce_scatter_ring.c

```diff
@@ -0,0 +1,364 @@
+    task->super.finalize = ucc_tl_ucp_reduce_scatter_ring_finalize;
+    task->subset         = subset;
+
+    if (task->subset.map.type != UCC_EP_MAP_FULL) {
+        status = ucc_ep_map_create_inverse(task->subset.map,
+                                           &task->reduce_scatter_ring.inv_map);
+        if (UCC_OK != status) {
+            return status;
+        }
+    } else {
+        task->reduce_scatter_ring.inv_map.type   = UCC_EP_MAP_FULL;
+        task->reduce_scatter_ring.inv_map.ep_num = task->subset.map.ep_num;
+    }
+    task->reduce_scatter_ring.n_frags         = n_frags;
+    task->reduce_scatter_ring.frag            = frag;
+    task->reduce_scatter_ring.scratch         = scratch;
+    task->reduce_scatter_ring.max_block_count = max_block_count;
+
+    *task_h = &task->super;
+    return UCC_OK;
+}
+
+static ucc_status_t
+ucc_tl_ucp_reduce_scatter_ring_sched_post(ucc_coll_task_t *coll_task)
+{
+    ucc_schedule_t *schedule = ucc_derived_of(coll_task, ucc_schedule_t);
+
+    return ucc_schedule_start(schedule);
+}
+
+static ucc_status_t
+ucc_tl_ucp_reduce_scatter_ring_sched_finalize(ucc_coll_task_t *task)
+{
+    ucc_tl_ucp_schedule_t *schedule = ucc_derived_of(task,
+                                                     ucc_tl_ucp_schedule_t);
+    ucc_status_t    status;
+
+    ucc_mc_free(schedule->scratch_mc_header);
+    status = ucc_schedule_finalize(task);
+    ucc_tl_ucp_put_schedule(&schedule->super.super);
+    return status;
+}
+
+ucc_status_t
+ucc_tl_ucp_reduce_scatter_ring_init(ucc_base_coll_args_t *coll_args,
+                                    ucc_base_team_t *     team,
+                                    ucc_coll_task_t **    task_h)
+{
+
+    ucc_tl_ucp_team_t *tl_team  = ucc_derived_of(team, ucc_tl_ucp_team_t);
+    ucc_rank_t         size     = UCC_TL_TEAM_SIZE(tl_team);
+    size_t             count    = coll_args->args.dst.info.count * size;
+    ucc_datatype_t     dt       = coll_args->args.dst.info.datatype;
+    size_t             dt_size  = ucc_dt_size(dt);
+    ucc_memory_type_t  mem_type = coll_args->args.dst.info.mem_type;
+    int                bidir =
+        UCC_TL_UCP_TEAM_LIB(tl_team)->cfg.reduce_scatter_ring_bidirectional;
+    size_t                 to_alloc_per_set, max_segcount, count_per_set;
+    ucc_tl_ucp_schedule_t *tl_schedule;
+    ucc_schedule_t        *schedule;
+    ucc_coll_task_t       *ctask;
+    ucc_status_t           status;
+    ucc_subset_t           s[2];
+    int                    i, n_subsets;
+
+    if (UCC_TL_UCP_TEAM_LIB(tl_team)->cfg.reduce_avg_pre_op &&
+        coll_args->args.op == UCC_OP_AVG) {
+        return UCC_ERR_NOT_SUPPORTED;
+    }
+
+    tl_schedule  = ucc_tl_ucp_get_schedule(tl_team, coll_args);
+    schedule     = &tl_schedule->super.super;
+    /* if count == size then we have 1 elem per rank, not enough
+       to split into 2 sets */
+    n_subsets    = (bidir && (count > size)) ? 2 : 1;
```

Comment:
does this condition work for inplace? i think should divide by size first in this case